File: public/dsa/linked_lists/remove_element_at_kth_position_in_linked_list/approach/remove_element_at_kth_position_in_linked_list.cpp

#include <bits/stdc++.h>
using namespace std;

// defs-start
class ListNode {
   public:
    int data;
    ListNode* next;
    ListNode(int data, ListNode* next = nullptr) : data(data), next(next) {}
};

// defs-end

ListNode* removekthElement(ListNode* head, int k) {
    // add your logic here
    if (k == 1) {
        return head->next;
    }
    auto temp = head;
    while (k > 2) {
        temp = temp->next;
        k--;
    }
    temp->next = temp->next->next;
    return head;
}

// tests-start
int main() {
    ListNode* head = new ListNode(1);
    head->next = new ListNode(3);
    head->next->next = new ListNode(4);
    head->next->next->next = new ListNode(7);
    int k = 2;
    auto res = removekthElement(head, k);
    auto temp = res;
    while (temp) {
        cout << temp->data << " ";
        temp = temp->next;
    }
    return 0;
}
// tests-end

/*output
1 4 7
*/

File: public/dsa/linked_lists/remove_element_at_kth_position_in_linked_list/approach/remove_element_at_kth_position_in_linked_list.py
# defs-start
class ListNode:
    def __init__(self, data=0, next=None):
        self.data = data
        self.next = next
# defs-end


class Solution:
    def removekthElement(self, head: ListNode, k: int) -> ListNode:
        if k == 1:
            return head.next
        temp = head
        while k > 2:
            temp = temp.next
            k -= 1
        temp.next = temp.next.next
        return head


# tests-start
head = ListNode(1, ListNode(3, ListNode(4, ListNode(7))))
sol = Solution()
res = sol.removekthElement(head, 2)

temp = res
while temp:
    print(temp.data, end=' ')
    temp = temp.next
# tests-end

'''output
1 4 7
'''
